From d234e3e8feefed596837f82acbb4e88f47a58143 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Mar 2019 07:40:58 +0100
Subject: [PATCH 29/44] ethtool: provide ring sizes in GET_PARAMS request

Add information about ring sizes (as provided by ETHTOOL_GRINGPARAM ioctl
command) in GET_PARAMS reply when ETH_PARAMS_IM_RING flag is set in the
request.

Send notification in the same format as reply when ring sizes are modified
using ioctl interface (ETHTOOL_SRINGPARAM command).

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt | 18 +++++-
 include/uapi/linux/ethtool_netlink.h         | 20 ++++++-
 net/ethtool/ioctl.c                          |  7 ++-
 net/ethtool/params.c                         | 59 ++++++++++++++++++++
 4 files changed, 99 insertions(+), 5 deletions(-)

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index 759c3f5313ea..30176c8410a0 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -410,8 +410,10 @@ values.
 GET_PARAMS
 ----------
 
-GET_PARAMS request retrieves information provided by ioctl comand
-ETHTOOL_GCOALESCE (coalescing parameters).
+GET_PARAMS request retrieves information provided by ioctl comands
+ETHTOOL_GCOALESCE (coalescing parameters) and ETHTOOL_GRINGPARAM (ring sizes).
+For each of these, there is a bit in header info_mask so that only one type of
+information can be requested.
 
 Request contents:
 
@@ -422,6 +424,7 @@ Request contents:
 Info mask bits:
 
     ETH_PARAMS_IM_COALESCE		coalescing parameters
+    ETH_PARAMS_IM_RING			ring sizes
 
 Response contents: On top level, there is one attribute for each of the
 information categories, the information is nested in it.
@@ -450,6 +453,15 @@ information categories, the information is nested in it.
         ETHA_COALESCE_TX_USE_ADAPTIVE		(bool)
         ETHA_COALESCE_RATE_SAMPLE_INTERVAL	(u32)
         ETHA_COALESCE_STATS_BLOCK_USECS		(u32)
+    ETHA_PARAMS_RING		(nested)	ring sizes
+        ETHA_RING_RX_MAX_PENDING		(u32)
+        ETHA_RING_RX_MINI_MAX_PENDING		(u32)
+        ETHA_RING_RX_JUMBO_MAX_PENDING		(u32)
+        ETHA_RING_TX_MAX_PENDING		(u32)
+        ETHA_RING_RX_PENDING			(u32)
+        ETHA_RING_RX_MINI_PENDING		(u32)
+        ETHA_RING_RX_JUMBO_PENDING		(u32)
+        ETHA_RING_TX_PENDING			(u32)
 
 GET_PARAMS requests allow dumps and messages in the same format as response
 to them are broadcasted as notifications on change of these settings using
@@ -511,7 +523,7 @@ ETHTOOL_GEEPROM			n/a
 ETHTOOL_SEEPROM			n/a
 ETHTOOL_GCOALESCE		ETHNL_CMD_GET_PARAMS
 ETHTOOL_SCOALESCE		ETHNL_CMD_SET_PARAMS
-ETHTOOL_GRINGPARAM		n/a
+ETHTOOL_GRINGPARAM		ETHNL_CMD_GET_PARAMS
 ETHTOOL_SRINGPARAM		n/a
 ETHTOOL_GPAUSEPARAM		n/a
 ETHTOOL_SPAUSEPARAM		n/a
diff --git a/include/uapi/linux/ethtool_netlink.h b/include/uapi/linux/ethtool_netlink.h
index 63b318a6a491..cc1aba2b3a11 100644
--- a/include/uapi/linux/ethtool_netlink.h
+++ b/include/uapi/linux/ethtool_netlink.h
@@ -293,14 +293,17 @@ enum {
 	ETHA_PARAMS_INFOMASK,			/* u32 */
 	ETHA_PARAMS_COMPACT,			/* flag */
 	ETHA_PARAMS_COALESCE,			/* nest - ETHA_COALESCE_* */
+	ETHA_PARAMS_RING,			/* nest - ETHA_RING_* */
 
 	__ETHA_PARAMS_CNT,
 	ETHA_PARAMS_MAX = (__ETHA_PARAMS_CNT - 1)
 };
 
 #define ETH_PARAMS_IM_COALESCE			(1U << 0)
+#define ETH_PARAMS_IM_RING			(1U << 1)
 
-#define ETH_PARAMS_IM_ALL (ETH_PARAMS_IM_COALESCE)
+#define ETH_PARAMS_IM_ALL (ETH_PARAMS_IM_COALESCE | \
+			   ETH_PARAMS_IM_RING)
 
 enum {
 	ETHA_COALESCE_UNSPEC,
@@ -331,6 +334,21 @@ enum {
 	ETHA_COALESCE_MAX = (__ETHA_COALESCE_CNT - 1)
 };
 
+enum {
+	ETHA_RING_UNSPEC,
+	ETHA_RING_RX_MAX_PENDING,		/* u32 */
+	ETHA_RING_RX_MINI_MAX_PENDING,		/* u32 */
+	ETHA_RING_RX_JUMBO_MAX_PENDING,		/* u32 */
+	ETHA_RING_TX_MAX_PENDING,		/* u32 */
+	ETHA_RING_RX_PENDING,			/* u32 */
+	ETHA_RING_RX_MINI_PENDING,		/* u32 */
+	ETHA_RING_RX_JUMBO_PENDING,		/* u32 */
+	ETHA_RING_TX_PENDING,			/* u32 */
+
+	__ETHA_RING_CNT,
+	ETHA_RING_MAX = (__ETHA_RING_CNT - 1)
+};
+
 /* generic netlink info */
 #define ETHTOOL_GENL_NAME "ethtool"
 #define ETHTOOL_GENL_VERSION 1
diff --git a/net/ethtool/ioctl.c b/net/ethtool/ioctl.c
index 911f597d632e..8496984237fa 100644
--- a/net/ethtool/ioctl.c
+++ b/net/ethtool/ioctl.c
@@ -1491,6 +1491,7 @@ static int ethtool_get_ringparam(struct net_device *dev, void __user *useraddr)
 static int ethtool_set_ringparam(struct net_device *dev, void __user *useraddr)
 {
 	struct ethtool_ringparam ringparam, max = { .cmd = ETHTOOL_GRINGPARAM };
+	int ret;
 
 	if (!dev->ethtool_ops->set_ringparam || !dev->ethtool_ops->get_ringparam)
 		return -EOPNOTSUPP;
@@ -1507,7 +1508,11 @@ static int ethtool_set_ringparam(struct net_device *dev, void __user *useraddr)
 	    ringparam.tx_pending > max.tx_max_pending)
 		return -EINVAL;
 
-	return dev->ethtool_ops->set_ringparam(dev, &ringparam);
+	ret = dev->ethtool_ops->set_ringparam(dev, &ringparam);
+	if (ret >= 0)
+		ethtool_notify(dev, NULL, ETHNL_CMD_SET_PARAMS,
+			       ETH_PARAMS_IM_RING, NULL);
+	return ret;
 }
 
 static noinline_for_stack int ethtool_get_channels(struct net_device *dev,
diff --git a/net/ethtool/params.c b/net/ethtool/params.c
index c653bf1a8d43..94f311e0e252 100644
--- a/net/ethtool/params.c
+++ b/net/ethtool/params.c
@@ -8,6 +8,7 @@ static const struct nla_policy get_params_policy[ETHA_PARAMS_MAX + 1] = {
 	[ETHA_PARAMS_INFOMASK]		= { .type = NLA_U32 },
 	[ETHA_PARAMS_COMPACT]		= { .type = NLA_FLAG },
 	[ETHA_PARAMS_COALESCE]		= { .type = NLA_REJECT },
+	[ETHA_PARAMS_RING]		= { .type = NLA_REJECT },
 };
 
 struct params_data {
@@ -16,6 +17,7 @@ struct params_data {
 	/* everything below here will be reset for each device in dumps */
 	struct common_reply_data	repdata_base;
 	struct ethtool_coalesce		coalesce;
+	struct ethtool_ringparam	ring;
 };
 
 static int parse_params(struct common_req_info *req_info, struct sk_buff *skb,
@@ -55,6 +57,15 @@ static int ethnl_get_coalesce(struct net_device *dev,
 	return dev->ethtool_ops->get_coalesce(dev, data);
 }
 
+static int ethnl_get_ring(struct net_device *dev,
+			  struct ethtool_ringparam *data)
+{
+	if (!dev->ethtool_ops->get_ringparam)
+		return -EOPNOTSUPP;
+	dev->ethtool_ops->get_ringparam(dev, data);
+	return 0;
+}
+
 static int prepare_params(struct common_req_info *req_info,
 			  struct genl_info *info)
 {
@@ -72,6 +83,11 @@ static int prepare_params(struct common_req_info *req_info,
 		if (ret < 0)
 			req_mask &= ~ETH_PARAMS_IM_COALESCE;
 	}
+	if (req_mask & ETH_PARAMS_IM_RING) {
+		ret = ethnl_get_ring(dev, &data->ring);
+		if (ret < 0)
+			req_mask &= ~ETH_PARAMS_IM_RING;
+	}
 	ethnl_after_ops(dev);
 
 	data->repdata_base.info_mask = req_mask;
@@ -86,6 +102,11 @@ static int coalesce_size(void)
 			      2 * nla_total_size(sizeof(u8)));
 }
 
+static int ring_size(void)
+{
+	return nla_total_size(8 * nla_total_size(sizeof(u32)));
+}
+
 static int params_size(const struct common_req_info *req_info)
 {
 	struct params_data *data =
@@ -96,6 +117,8 @@ static int params_size(const struct common_req_info *req_info)
 	len += dev_ident_size();
 	if (info_mask & ETH_PARAMS_IM_COALESCE)
 		len += coalesce_size();
+	if (info_mask & ETH_PARAMS_IM_RING)
+		len += ring_size();
 
 	return len;
 }
@@ -158,6 +181,36 @@ static int fill_coalesce(struct sk_buff *skb, struct ethtool_coalesce *data)
 	return 0;
 }
 
+static int fill_ring(struct sk_buff *skb, struct ethtool_ringparam *data)
+{
+	struct nlattr *nest = ethnl_nest_start(skb, ETHA_PARAMS_RING);
+
+	if (!nest)
+		return -EMSGSIZE;
+	if (nla_put_u32(skb, ETHA_RING_RX_MAX_PENDING,
+			data->rx_max_pending) ||
+	    nla_put_u32(skb, ETHA_RING_RX_MINI_MAX_PENDING,
+			data->rx_mini_max_pending) ||
+	    nla_put_u32(skb, ETHA_RING_RX_JUMBO_MAX_PENDING,
+			data->rx_jumbo_max_pending) ||
+	    nla_put_u32(skb, ETHA_RING_TX_MAX_PENDING,
+			data->tx_max_pending) ||
+	    nla_put_u32(skb, ETHA_RING_RX_PENDING,
+			data->rx_pending) ||
+	    nla_put_u32(skb, ETHA_RING_RX_MINI_PENDING,
+			data->rx_mini_pending) ||
+	    nla_put_u32(skb, ETHA_RING_RX_JUMBO_PENDING,
+			data->rx_jumbo_pending) ||
+	    nla_put_u32(skb, ETHA_RING_TX_PENDING,
+			data->tx_pending)) {
+		nla_nest_cancel(skb, nest);
+		return -EMSGSIZE;
+	}
+
+	nla_nest_end(skb, nest);
+	return 0;
+}
+
 static int fill_params(struct sk_buff *skb,
 		       const struct common_req_info *req_info)
 {
@@ -171,6 +224,11 @@ static int fill_params(struct sk_buff *skb,
 		if (ret < 0)
 			return ret;
 	}
+	if (info_mask & ETH_PARAMS_IM_RING) {
+		ret = fill_ring(skb, &data->ring);
+		if (ret < 0)
+			return ret;
+	}
 
 	return 0;
 }
@@ -196,6 +254,7 @@ static const struct nla_policy set_params_policy[ETHA_PARAMS_MAX + 1] = {
 	[ETHA_PARAMS_INFOMASK]		= { .type = NLA_REJECT },
 	[ETHA_PARAMS_COMPACT]		= { .type = NLA_FLAG },
 	[ETHA_PARAMS_COALESCE]		= { .type = NLA_NESTED },
+	[ETHA_PARAMS_RING]		= { .type = NLA_REJECT },
 };
 
 static const struct nla_policy coalesce_policy[ETHA_COALESCE_MAX + 1] = {
-- 
2.21.0

